# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
def addTwo(x):
    return x+2
print(addTwo(4))
#Lambda function syntax
a_var = 6
print((lambda x: x+2)(a_var))

6
8


#### Saving to a Variable

In [2]:
f_test = lambda x: x+2
f_test (4)

6

#### Multiple Inputs

In [4]:
#multiple input with no variable
print((lambda x, y, z : x * y * z)(3, 4, 5))
x_test = lambda x, y ,z : x * y * z
print(x_test(2, 5, 8))

60
80


#### Passing a Lambda into a Function

In [5]:
def multiply(f , num):
    """f expects a lambda function
    num expects a number"""
    return f(num)

multiply(lambda x : x * x, 4)





16

#### Returning a Lambda from a Function

In [13]:
#Normal defined function
def multiply_test(num):
    return num * 4
#Function within the function
def returnFunc():
    test = 4
    def multiply(num):
        return num * 3
    return multiply
f_return = returnFunc()
print(returnFunc())
print(f_return(5))

#Lambda function return from a regular function
def return_lam(b, c):
    return lambda x, a: x + a + b + c
r_lamb = return_lam(4,8)
print(r_lamb(4,4)) 

<function returnFunc.<locals>.multiply at 0x000002E6E4F06660>
15
20


#### If Statements within Lambdas

In [14]:
#Lambda x: True if(condition) else False
f_condition = lambda num: num * 2 if num > 10 else num + 2
print(f_condition(8))
print(f_condition(10))

10
12


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [28]:
def cube_test(f, num):
    return f(num)
cube_test(lambda x : x ** 3,  10)
#cube_test(lambda x : x * x * x, 5)

 


1000

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [29]:
#Map(function, iterable)
# an iterable is something you can loop over: such as list, dict, tuple set or string
#Normally the usage of Map happens with a predefined function but we can use lambda as well
def squared (num1 ,num2):
    if num1 < 10 and num2 < 10:
        return num1 ** 2, num2 ** 2
    else:
        return num1, num2
numbers = [2, 5, 8, 13, 10, 20]
more_numbers = [4, 10, 3, 2, 6]
squared_num_map = list(map(squared, numbers, more_numbers)) 
print(squared_num_map)



[(4, 16), (5, 10), (64, 9), (13, 2), (10, 6)]


#### Using Lambda's with Map

In [45]:
squared_num_lamb = list(map(lambda x, y:( x ** 2, y ** 2) if  x < 10 and y < 10 else (x, y), numbers, more_numbers))
print(squared_num_lamb)

[(1, 16), (2, 10), (9, 9), (16, 4), (25, 36)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [46]:
double_minus_lamb = list (map(lambda x ,y:(2*x - 1, 2*y- 1) if x < 10 and y < 10 else (x, y), numbers, more_numbers))
print (double_minus_lamb )

# Define a list of numbers
numbers = [1, 2, 3, 4, 5]

# Use map() with a lambda function to double each number and subtract 1
result = list(map(lambda x: 2*x - 1, numbers))

# Print the result
print(result)

[(1, 7), (2, 10), (5, 5), (7, 3), (9, 11)]
[1, 3, 5, 7, 9]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [50]:
names = ["Bob", "Andy", "Max", "Evan", "Angelica"]
def a_names (name):
    if name[0].lower() == 'a':
        return True
    else:
        return False
new_names_list = list(filter(a_names, names))
print(new_names_list)     


['Andy', 'Angelica']


#### Using Lambda's with Filter()

In [51]:
new_names_lamb = list(filter(lambda name : True if name[0].lower() == 'a' else False, names))
print(new_names_lamb)

['Andy', 'Angelica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [58]:
import statistics

# Define a list of numbers
numbers = [2, 7, 4.2, 1.6, 9, 4.4, 4.9]

# Calculate the mean of the list
mean = statistics.mean(numbers)
print(mean)

# Use filter() to filter out all the numbers below the mean
result = list(filter(lambda x: x >= mean, numbers))
# Print the result
print(result)




4.728571428571429
[7, 9, 4.9]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

#### Using Lambda's with Reduce()

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [48]:
from functools import reduce

# Define a list of numbers
numbers = [1, 2, 3, 4, 5]

# Use reduce() with a lambda function to multiply the numbers together
result = reduce(lambda x, y: x * y, numbers)

# Print the result
print(result)


120


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [61]:
def addNums(num):
    #set a base case to stop the recursion  from going forever
    if num <= 1:
        print("addNums(1) = 1")
        return num
    else:
        print(f"addNums({num} = {num} + addNums({num-1}))")
        return num +addNums(num-1)
addNums(5)

addNums(5 = 5 + addNums(4))
addNums(4 = 4 + addNums(3))
addNums(3 = 3 + addNums(2))
addNums(2 = 2 + addNums(1))
addNums(1) = 1


15

#### Writing a Factorial Function

In [62]:
# 5 ! = 5 * 4 * 3 * 2 * 1
def factorial(num):
    if num <= 1:
        return 1
    else:
        return num * factorial(num - 1)
factorial(5)   

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [66]:
def subtract_numbers(num):
    if num == 0:
        return 0
    else:
        return num - subtract_numbers(num-1)
result = subtract_numbers(5)
print(result)

def subNums(num):
    #set a base case to stop the recursion  from going forever
    if num <= 1:
        print("subNums(1) = 1")
        return num
    else:
        print(f"subNums({num} = {num} - subNums({num-1}))")
        return num -subNums(num-1)
subNums(5)


3
subNums(5 = 5 - subNums(4))
subNums(4 = 4 - subNums(3))
subNums(3 = 3 - subNums(2))
subNums(2 = 2 - subNums(1))
subNums(1) = 1


3

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [67]:
def my_range(stop, start = 0, step = 2):
    while start < stop :
        yield start
        start += step
for i in my_range(20):
    my_generator_value = i
    print(my_generator_value)       
         

0
2
4
6
8
10
12
14
16
18


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [5]:
def gen_squares(num):
    while num > 0:
        square = num ** 2
        yield square
        print(square)
        num -= 1
for squared in gen_squares(5):
    pass



25
16
9
4
1


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [6]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]
filtered_places = list(filter(lambda x: x.strip() != "", places))
print(filtered_places)


['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [10]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author.sort(key=lambda x: x.split()[-1].lower())
print(author)


['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [9]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
fahrenheit_places = list(map(lambda x: (x[0], (9/5)*x[1]+32), places))
print(fahrenheit_places)




[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [3]:
def fib(n):
    if n == 0 or n == 1:
        return 1
    else:
        return fib(n-1) + fib(n-2)

def print_fib_sequence(n):
    for i in range(n+1):
        print("Iteration {}: {}".format(i, fib(i)))

print_fib_sequence(5)


Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8
